pkg/virtctl/create/vm.go

```diff
@@ -0,0 +1,321 @@
+
+  # Create a VirtualMachine with a specified Instancetype, Preference and image source in other namespace
+  {{ProgramName}} create vm --instancetype=my-instancetype --preference=my-preference --image-source=pvc/my-pvc.my-namespace
+  
+  # Create a VirtualMachine with a specified ClusterInstancetype, ClusterPreference and direct image source
+  {{ProgramName}} create vm --clusterinstancetype=my-clusterinstancetype --clusterpreference=my-clusterpreference --direct-image-source=dv/my-dv`
+}
+
+func newVM(c *createVM) *v1.VirtualMachine {
+	runStrategy := v1.VirtualMachineRunStrategy(c.runStrategy)
+	return &v1.VirtualMachine{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       v1.VirtualMachineGroupVersionKind.Kind,
+			APIVersion: v1.VirtualMachineGroupVersionKind.GroupVersion().String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: c.name,
+		},
+		Spec: v1.VirtualMachineSpec{
+			RunStrategy: &runStrategy,
+			Template: &v1.VirtualMachineInstanceTemplateSpec{
+				Spec: v1.VirtualMachineInstanceSpec{
+					TerminationGracePeriodSeconds: &c.terminationGracePeriod,
+				},
+			},
+		},
+	}
+}
+
+func splitKindName(kindName string) (string, string, error) {
+	split := strings.Split(kindName, "/")
+	switch l := len(split); l {
+	case 1:
+		return "", split[0], nil
+	case 2:
+		return strings.ToLower(split[0]), split[1], nil
+	default:
+		return "", "", fmt.Errorf("Invalid count %d of slashes in kind/name", l)
+	}
+}
+
+func withInstancetype(c *createVM, vm *v1.VirtualMachine) error {
+	kind, name, err := splitKindName(c.instancetype)
+	if err != nil {
+		return err
+	}
+
+	if kind != "" && kind != instancetype.SingularResourceName && kind != instancetype.ClusterSingularResourceName {
+		return fmt.Errorf("Invalid instancetype kind %s", kind)
+	}
+
+	vm.Spec.Instancetype = &v1.InstancetypeMatcher{
+		Name: name,
+		Kind: kind,
+	}
+
+	return nil
+}
+
+func withPreference(c *createVM, vm *v1.VirtualMachine) error {
+	kind, name, err := splitKindName(c.preference)
+	if err != nil {
+		return err
+	}
+
+	if kind != "" && kind != instancetype.SingularPreferenceResourceName && kind != instancetype.ClusterSingularPreferenceResourceName {
+		return fmt.Errorf("Invalid preference kind %s", kind)
+	}
+
+	vm.Spec.Preference = &v1.PreferenceMatcher{
+		Name: name,
+		Kind: kind,
+	}
+
+	return nil
+}
+
+func withImageSource(c *createVM, vm *v1.VirtualMachine) error {
+	kind, name, namespace, err := parseImageSource(c.imageSource)
+	if err != nil {
+		return err
+	}
+
+	vm.Spec.DataVolumeTemplates = []v1.DataVolumeTemplateSpec{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: name,
+			},
+			Spec: cdiv1.DataVolumeSpec{
+				Storage: &cdiv1.StorageSpec{},
+				SourceRef: &cdiv1.DataVolumeSourceRef{
+					Kind: kind,
+					Name: name,
+				},
+			},
+		},
+	}
+	if namespace != "" {
+		vm.Spec.DataVolumeTemplates[0].Spec.SourceRef.Namespace = &namespace
+	}
+
+	vm.Spec.Template.Spec.Volumes = []v1.Volume{
```

Comment:
I assumed we'd be chaining these switches togther to allow multiple volumes to be defined?

Reply:
How to determine the boot order? The order in which the switches were defined?

Reply:
Yes, the ordering is then retained if we generate a list of disks in the VMI mutation webhook.

Reply:
There are four switches, their order is:

```
--volume-dvt-ds
--volume-dvt-pvc
--volume-dv
--volume-pvc  
```

Each switch can be specified multiple times, the order in which it was specified will be kept.

Example 1:

```
virtctl create vm --volume-dvt-ds my-ds1,my-ds2 --volume-dv my-dv1 --volume-dv my-dv2
-->
vm.Spec.Template.Spec.Volumes = my-ds1,my-ds2,my-dv1,my-dv2

virtctl create vm --volume-dv my-dv1 --volume-dv my-dv2 --volume-dvt-pvc my-ns/my-pvc
-->
vm.Spec.Template.Spec.Volumes = my-ns/my-pvc,my-dv1,my-dv2
```